New option to control adding to the jumplist

lag13 opened the pull request:
Hello again! Merry Christmas!! Here's a present? This PR adds a new option which solves a slight annoyance I've encountered while editing. I'll illustrate with an example. Say I'm happily editing a file and I want to change a string somewhere in that file. So I'll do a sequence of commands like this:

```vim
" Go to the string I want to edit, this leaves my cursor on the opening quote of the string
/'str<CR>
" Changes the contents of the string
ci'hello world<ESC>
``` 

After doing that I might want to go back to where I was before the search and so I instinctively do ``` `` ```, but it doesn't quite work. Since this plugin adds the cursor position before the selection of the text object to the jump list, doing ``` `` ``` will actually just move me to the opening quote of the string which now says `'hello world'`. To get back to where I was originally, I'll need to issue a couple `<C-o>`'s. Not terrible by any means (I can't honestly say how often it comes up. I've been meaning to tally up when it happens and get a basic sense of frequency) but I thought it would be nice if it could be fixed! Had some free time this vacation so I did.

### Your Opinion

Do you think this a useful option to add? Is the sort of problem I described ever something you encounter? The short summary of possible settings for this new option are:

1. Always add to the jump list (same as current behavior)
2. Only add to the jump list if the cursor position prior to selection was not inside the 'around'/'raw' version of the text object. In doing this I wondered if I should add another setting similar to this where we add to the jump list if the cursor was not inside the *actual* text object being selected rather than the 'raw' text object. If this option existed and was set then doing `ci'` in my above example *would* have added the cursor position to the jump list because my cursor was on the opening `'` which is not contained by the `i'` text object. But this wasn't the behavior I was interested in so I decided not to implement it. But it would be simple to add that behavior in if you think someone might want or just for the heck of it.
3. Never add to the jump list. Honestly I don't know if anyone would want this. I kind of just threw it in for the sake of completeness (because then you get on, off, and half on).

### Found a couple bugs!

In the process of making this PR I found a couple things which I think are bugs. This plugin adds to the jump list multiple times when the 'argument' or 'pair' text objects get used.

#### https://github.com/wellle/targets.vim/blob/master/autoload/targets.vim#L711

I wasn't aware of this but it seems to be the case that in vanilla vim when you use a text object described in the help pages as a "block" (like `i(` or `a{`) then it adds the cursor position prior to the selection to the jump list. So whenever this line of code runs it will add to the jump list.

#### https://github.com/wellle/targets.vim/blob/master/autoload/targets.vim#L862

Whenever you have to skip over a function call, this line of code will add to the jump list.

### Yep

So yeah, I think that's about it. Merry christmas!

wellle commented:
Hey @lag13, thanks and merry christmas to you too!

This looks like a solid contribution and I like the idea of using keepjumps and adding to the jump list explicitly.

In general I think we should use more readable values instead of magic numbers for such a setting.

I would like most if we could agree on one set of behaviors that should be fine for everyone instead of introducing a setting. Do you think that would be feasable? In general I believe we should only add to the jumplist on a "long jump". For example we could consider to only add if the cursor was on a line that is not contained in the line that intersect the target. I believe this could be implemented based on the seeking implementation. You might have a look at the cheat sheet for details. 

I'll try to have a deeper look in the next days. Thanks for contributing!

lag13 commented:
All your suggestions sound good to me:

- If we can avoid adding a setting then great. I had introduced one strictly for the sake of maintaining backwards compatibility.
- If we decided to make such a setting we'd use more obvious values (like strings) instead of magic numbers.
- Pick whatever behavior makes the most sense to you. It seems like the 2 possible implementations that could work would be to add to the jumplist if the cursor was outside of the 'raw' text object (the behavior I implemented) or if it was on a different line than the text object (your suggestion). As long as it satisfies my use case I don't care overly much.

As always take your time to check this out, there is no rush whatsoever. Also, is this sort of behavior something that makes sense to you to add? Would you find it useful? Because I wouldn't want to add something potentially unnecessary.

wellle commented:
This is definitely something I care about. But feel free to stop working on it until I make up my mind :)

wellle commented:
Included in #139. Thanks!
